civiform/civiform
# Save uploaded file when clicking Review or Previous [behind SAVE_ON_ALL_ACTIONS flag]

### Description

To address https://github.com/civiform/civiform/issues/6450, we'd like to save any user data whenever *any* button is clicked in an application. Previous PRs have made the "Review" and "Previous" buttons save data for the typical block:
 - Review button: https://github.com/civiform/civiform/pull/6544
 - Previous button: https://github.com/civiform/civiform/pull/6547

However, those PRs didn't update the "Review" and "Previous" buttons for the file upload block because the file upload block has some custom logic. This PR updates the file upload block to save data when "Review" and "Previous" buttons are clicked.

### Background Details

 #### How saving on all actions works

Previously, a user's answers were only saved when a user clicked "Save&next". So, the "Save&next" button invoked an endpoint called `ApplicantProgramBlocksController#update`, which would (1) Save the user's data to the database; and (2) Redirect the user to the next block in the program.

To have *all* buttons save the user's data, we've updated all buttons to use that `ApplicantProgramBlocksController#update` endpoint. We've added a parameter to the endpoint called `applicantRequestedAction` which indicates *where* the redirect should go ("Save&next" passes in `NEXT_BLOCK`, "Previous" passes in `PREVIOUS_BLOCK`, "Review" passes in `REVIEW_PAGE`.)

 #### How file upload works

When a user uploads a file then clicks "Save & next", the HTML form is submitted to AWS (or other cloud storage provider), *not* CiviForm. AWS stores the file in a bucket, and if the file was successfully stored then AWS invokes the URL specified by the `<input name="success_action_redirect">` in the HTML form. We currently set the `success_action_redirect` URL to point to `ApplicantProgramBlocksController#updateFile`. That endpoint will (1) Save the file key in the database so that we know how to fetch the file back from AWS later; and (2) Redirect the user to the next block in the program.

Now that all the buttons should save the file, all the buttons will need to follow a similar pattern of submitting the HTML form to AWS First. They'll also need the `#updateFile` endpoint to handle redirecting to different places.

#### File upload question requirements

Because the file upload question first sends the HTML form to AWS, the file upload question can't be on a block with any other questions because AWS will throw away all the other question inputs. File upload questions are always on a block by themselves.

### Expected behaviors

See the manual testing directions for expected behaviors for each of the buttons.

### Implementation

Given those background details above, this PR makes a few changes:
1) Updates all buttons on the file upload question page to be `submitButton`s so that they send the HTML form to AWS first, then trigger `ApplicantProgramBlocksController#updateFile`.
1) Updates `ApplicantProgramBlocksController#updateFile` to also take in an `applicantRequestedAction` parameter so that it can redirect to different places based on which button was clicked.
3) Updates the TypeScript on the file upload question page to change what the `<input name="success_action_redirect"> is based on which button was clicked.
4) Updates the TypeScript on the file upload question page to allow the "Review" and "Previous" buttons to do the correct redirect even if the user didn't upload a file.

## Release notes

### Checklist

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)


#### User visible changes


#### New Features


### Instructions for manual testing

Notably: The Previous and Review buttons work correctly whether or not the user uploaded a file. The Save&next button will only work if the user *has* uploaded a file.

Basic buttons:
 - Save&next:
     -  *without* an uploaded file -> an error is shown and the user stays on the block. **Note:** This is true for both required file upload questions and optional file upload questions. Optional file upload questions include a separate Skip button to let the user proceed to the next block without uploading a file.
    - *with* an uploaded file -> file saved, user taken to next block
 - Previous:
      -  *without* an uploaded file -> user taken to previous block
      - *with* an uploaded file -> filed saved, user taken to previous block
 - Review:
      - *without* an uploaded file -> user taken to review page
     - *with* an uploaded file -> filed saved, user taken to review page
 
Bonus buttons:
 - Continue (only shown if you have previously uploaded a file then come back to the question):
      -  *without* an uploaded file -> user taken to next block
      - *with* an uploaded file -> file **not** saved, user taken to next block. **Note:** I think this behavior is worth changing, but that's out-of-scope for this PR.
 - Skip (only shown on optional file upload questions that don't already have a previously uploaded file):
      -  *without* an uploaded file -> user taken to next block
      - *with* an uploaded file -> file **not** saved, user taken to next block
 - Delete (only shown on optional file upload questions and if you have previously uploaded a file then come back to the question):
     - deletes previously uploaded file, user taken to next block

### Issue(s) this completes

Part of #6450
